depends: don't restrict --enable-lto to non-guix cctools

fanquake opened the pull request:
This wasn't actually disabling LTO support anyways, because it's enabled
by default.

Guix Build (x86_64):
```bash
af6e0b5fd494f714a84716161352aee415700c1f50f1f08665a474dc2ed56bad  guix-build-7f73f422a49f/output/aarch64-linux-gnu/SHA256SUMS.part
1ecdbf2b579204cf2086001df3887da2ba50cd3f7b8f735113cce8f1b634fdb2  guix-build-7f73f422a49f/output/aarch64-linux-gnu/bitcoin-7f73f422a49f-aarch64-linux-gnu-debug.tar.gz
878ea08855dc559e62fd1932f8a54d53fc63e0ce7d587345c7326892622feee9  guix-build-7f73f422a49f/output/aarch64-linux-gnu/bitcoin-7f73f422a49f-aarch64-linux-gnu.tar.gz
4d1d35bd329935bc74b82a78a7fb8dd39c3268d4bbb33ecd877079e49de97a88  guix-build-7f73f422a49f/output/arm-linux-gnueabihf/SHA256SUMS.part
c0c5418dc270d716b354d2190fddde3d25794758dee45c9068e9f5aac20e5617  guix-build-7f73f422a49f/output/arm-linux-gnueabihf/bitcoin-7f73f422a49f-arm-linux-gnueabihf-debug.tar.gz
5e13ac8cacf1042746eef7e85695469c50d6afc06dfdb5da53039175e5185b52  guix-build-7f73f422a49f/output/arm-linux-gnueabihf/bitcoin-7f73f422a49f-arm-linux-gnueabihf.tar.gz
218b2463a07503e77711fcf0245bc642eb9ed76e4a3e8b61d66a5c76bdc25c62  guix-build-7f73f422a49f/output/arm64-apple-darwin/SHA256SUMS.part
dd15af2b7a333f3c4b3f75abd307e1789a1629728a0ae33d8a509478ace91134  guix-build-7f73f422a49f/output/arm64-apple-darwin/bitcoin-7f73f422a49f-arm64-apple-darwin-unsigned.dmg
5cdfd9fc347007890991c68746b1bb36e0cced57fee685337b3d079d498f687d  guix-build-7f73f422a49f/output/arm64-apple-darwin/bitcoin-7f73f422a49f-arm64-apple-darwin-unsigned.tar.gz
80f957e78177927ace323ea90d4bad2e8d2b846b199a0dcb085c3ceae8516631  guix-build-7f73f422a49f/output/arm64-apple-darwin/bitcoin-7f73f422a49f-arm64-apple-darwin.tar.gz
3cb9cb04db3ebf328bfabfda874fce09764f3840c98bd850d85f71892617fb99  guix-build-7f73f422a49f/output/dist-archive/bitcoin-7f73f422a49f.tar.gz
e4e7c9d5366b658fc420c1fd3cf12bfb9863b8061c83b06fe2492e934bb75ae9  guix-build-7f73f422a49f/output/powerpc64-linux-gnu/SHA256SUMS.part
da4ea7dd9a131901ff2f7fcd6eba573fe946ea671ae6ada418a4ceebcc4bd3b6  guix-build-7f73f422a49f/output/powerpc64-linux-gnu/bitcoin-7f73f422a49f-powerpc64-linux-gnu-debug.tar.gz
221f6e1d5193207eed400c186c8f83747fbf90c3244cd044b36acadbed0c83a6  guix-build-7f73f422a49f/output/powerpc64-linux-gnu/bitcoin-7f73f422a49f-powerpc64-linux-gnu.tar.gz
12d36079b330ec75209219ce58ca580ca5a5eeac64fb7ae424c64e7f3d466ee5  guix-build-7f73f422a49f/output/powerpc64le-linux-gnu/SHA256SUMS.part
3cef0194569d6fa968e92be1d2189be4f1337272dc213c25b8d4afe6b174310c  guix-build-7f73f422a49f/output/powerpc64le-linux-gnu/bitcoin-7f73f422a49f-powerpc64le-linux-gnu-debug.tar.gz
a8ff109153926b3240f830c3501977046b6c470b67380a01c55aea2401a1a833  guix-build-7f73f422a49f/output/powerpc64le-linux-gnu/bitcoin-7f73f422a49f-powerpc64le-linux-gnu.tar.gz
70b3a841bb232fefe27f857ee826ea6a84f59bc781acda608bd6d9385a1ef149  guix-build-7f73f422a49f/output/riscv64-linux-gnu/SHA256SUMS.part
841a276263877027f256c3b7ad1c5bc3ae3ad4d6a7157690daa947f6809776e2  guix-build-7f73f422a49f/output/riscv64-linux-gnu/bitcoin-7f73f422a49f-riscv64-linux-gnu-debug.tar.gz
6cce2c36c2a53035d7ee5d285c8b4b99256772c7a9dca1739f0639ecc0bed71d  guix-build-7f73f422a49f/output/riscv64-linux-gnu/bitcoin-7f73f422a49f-riscv64-linux-gnu.tar.gz
e910b32efebca682cf8bc3994d2c7513b16cdeb6f94cef906efb6629ec53268d  guix-build-7f73f422a49f/output/x86_64-apple-darwin/SHA256SUMS.part
3a4bc1573d9c576184d69b953ea6a85650d4e15b71503b909ccd4fb8a69ca383  guix-build-7f73f422a49f/output/x86_64-apple-darwin/bitcoin-7f73f422a49f-x86_64-apple-darwin-unsigned.dmg
60a114ece1d9a274d9f3d80a9633eb22d589213ba53a956f3b1138d7c8f55560  guix-build-7f73f422a49f/output/x86_64-apple-darwin/bitcoin-7f73f422a49f-x86_64-apple-darwin-unsigned.tar.gz
b6c6eca3e81bfadfed3ddcb7830857c954c8a7b0ff458f14050f4952240ebf27  guix-build-7f73f422a49f/output/x86_64-apple-darwin/bitcoin-7f73f422a49f-x86_64-apple-darwin.tar.gz
aab972c6f051c379a20eb694128a4df4e3737758fbca3902d28d72cd93e6d33a  guix-build-7f73f422a49f/output/x86_64-linux-gnu/SHA256SUMS.part
b8e643eeb72cf04664462f815497cbef1443f2e113de14bc93b058d83836c7df  guix-build-7f73f422a49f/output/x86_64-linux-gnu/bitcoin-7f73f422a49f-x86_64-linux-gnu-debug.tar.gz
3b40fb5f1341cc3b49c9eccc68fc75e1cc7e7d2c6a5664fa2f8f095290e3733a  guix-build-7f73f422a49f/output/x86_64-linux-gnu/bitcoin-7f73f422a49f-x86_64-linux-gnu.tar.gz
96138d5f591a8ba8589eb81d2c79ee269556996ae93c6a3fc77746e101b60cab  guix-build-7f73f422a49f/output/x86_64-w64-mingw32/SHA256SUMS.part
6ef2d9436caa550bec7defda1d769bffc2e0fe8f7acf99e2fbcefd2e9b920a64  guix-build-7f73f422a49f/output/x86_64-w64-mingw32/bitcoin-7f73f422a49f-win64-debug.zip
91c8c3f0aec623f4183c4ebc141816dddb4482509e99812d32298f3e83da920f  guix-build-7f73f422a49f/output/x86_64-w64-mingw32/bitcoin-7f73f422a49f-win64-setup-unsigned.exe
1de7cd32d74990cfcd4e00428c53a35742b7a480f18ae1bd60dbf91b60400947  guix-build-7f73f422a49f/output/x86_64-w64-mingw32/bitcoin-7f73f422a49f-win64-unsigned.tar.gz
e4d4ba33bbdd8d9181c8422af31ad6127f3627c327a93cbe8245d7154b5a936c  guix-build-7f73f422a49f/output/x86_64-w64-mingw32/bitcoin-7f73f422a49f-win64.zip
```

Guix Build (arm64):
```bash
4d1d35bd329935bc74b82a78a7fb8dd39c3268d4bbb33ecd877079e49de97a88  guix-build-7f73f422a49f/output/arm-linux-gnueabihf/SHA256SUMS.part
c0c5418dc270d716b354d2190fddde3d25794758dee45c9068e9f5aac20e5617  guix-build-7f73f422a49f/output/arm-linux-gnueabihf/bitcoin-7f73f422a49f-arm-linux-gnueabihf-debug.tar.gz
5e13ac8cacf1042746eef7e85695469c50d6afc06dfdb5da53039175e5185b52  guix-build-7f73f422a49f/output/arm-linux-gnueabihf/bitcoin-7f73f422a49f-arm-linux-gnueabihf.tar.gz
5768bc05947954f8d0caa1869edbba7a9102339d43c47a43cbc5eabdabebfd16  guix-build-7f73f422a49f/output/arm64-apple-darwin/SHA256SUMS.part
e506d3a635364751eaeccf553bd8c8847e5a2ab4a28bb03108052ae697b49e7e  guix-build-7f73f422a49f/output/arm64-apple-darwin/bitcoin-7f73f422a49f-arm64-apple-darwin-unsigned.dmg
76fd8b6954ead8bbbba56fbf747459019dcd2def22ab6615ce817f2fff7adcb8  guix-build-7f73f422a49f/output/arm64-apple-darwin/bitcoin-7f73f422a49f-arm64-apple-darwin-unsigned.tar.gz
f369a59d44cd31a505faf1eefd84b9ecde91f4cb036e79d38188c9ed0f9c094c  guix-build-7f73f422a49f/output/arm64-apple-darwin/bitcoin-7f73f422a49f-arm64-apple-darwin.tar.gz
3cb9cb04db3ebf328bfabfda874fce09764f3840c98bd850d85f71892617fb99  guix-build-7f73f422a49f/output/dist-archive/bitcoin-7f73f422a49f.tar.gz
e4e7c9d5366b658fc420c1fd3cf12bfb9863b8061c83b06fe2492e934bb75ae9  guix-build-7f73f422a49f/output/powerpc64-linux-gnu/SHA256SUMS.part
da4ea7dd9a131901ff2f7fcd6eba573fe946ea671ae6ada418a4ceebcc4bd3b6  guix-build-7f73f422a49f/output/powerpc64-linux-gnu/bitcoin-7f73f422a49f-powerpc64-linux-gnu-debug.tar.gz
221f6e1d5193207eed400c186c8f83747fbf90c3244cd044b36acadbed0c83a6  guix-build-7f73f422a49f/output/powerpc64-linux-gnu/bitcoin-7f73f422a49f-powerpc64-linux-gnu.tar.gz
12d36079b330ec75209219ce58ca580ca5a5eeac64fb7ae424c64e7f3d466ee5  guix-build-7f73f422a49f/output/powerpc64le-linux-gnu/SHA256SUMS.part
3cef0194569d6fa968e92be1d2189be4f1337272dc213c25b8d4afe6b174310c  guix-build-7f73f422a49f/output/powerpc64le-linux-gnu/bitcoin-7f73f422a49f-powerpc64le-linux-gnu-debug.tar.gz
a8ff109153926b3240f830c3501977046b6c470b67380a01c55aea2401a1a833  guix-build-7f73f422a49f/output/powerpc64le-linux-gnu/bitcoin-7f73f422a49f-powerpc64le-linux-gnu.tar.gz
70b3a841bb232fefe27f857ee826ea6a84f59bc781acda608bd6d9385a1ef149  guix-build-7f73f422a49f/output/riscv64-linux-gnu/SHA256SUMS.part
841a276263877027f256c3b7ad1c5bc3ae3ad4d6a7157690daa947f6809776e2  guix-build-7f73f422a49f/output/riscv64-linux-gnu/bitcoin-7f73f422a49f-riscv64-linux-gnu-debug.tar.gz
6cce2c36c2a53035d7ee5d285c8b4b99256772c7a9dca1739f0639ecc0bed71d  guix-build-7f73f422a49f/output/riscv64-linux-gnu/bitcoin-7f73f422a49f-riscv64-linux-gnu.tar.gz
e910b32efebca682cf8bc3994d2c7513b16cdeb6f94cef906efb6629ec53268d  guix-build-7f73f422a49f/output/x86_64-apple-darwin/SHA256SUMS.part
3a4bc1573d9c576184d69b953ea6a85650d4e15b71503b909ccd4fb8a69ca383  guix-build-7f73f422a49f/output/x86_64-apple-darwin/bitcoin-7f73f422a49f-x86_64-apple-darwin-unsigned.dmg
60a114ece1d9a274d9f3d80a9633eb22d589213ba53a956f3b1138d7c8f55560  guix-build-7f73f422a49f/output/x86_64-apple-darwin/bitcoin-7f73f422a49f-x86_64-apple-darwin-unsigned.tar.gz
b6c6eca3e81bfadfed3ddcb7830857c954c8a7b0ff458f14050f4952240ebf27  guix-build-7f73f422a49f/output/x86_64-apple-darwin/bitcoin-7f73f422a49f-x86_64-apple-darwin.tar.gz
aab972c6f051c379a20eb694128a4df4e3737758fbca3902d28d72cd93e6d33a  guix-build-7f73f422a49f/output/x86_64-linux-gnu/SHA256SUMS.part
b8e643eeb72cf04664462f815497cbef1443f2e113de14bc93b058d83836c7df  guix-build-7f73f422a49f/output/x86_64-linux-gnu/bitcoin-7f73f422a49f-x86_64-linux-gnu-debug.tar.gz
3b40fb5f1341cc3b49c9eccc68fc75e1cc7e7d2c6a5664fa2f8f095290e3733a  guix-build-7f73f422a49f/output/x86_64-linux-gnu/bitcoin-7f73f422a49f-x86_64-linux-gnu.tar.gz
96138d5f591a8ba8589eb81d2c79ee269556996ae93c6a3fc77746e101b60cab  guix-build-7f73f422a49f/output/x86_64-w64-mingw32/SHA256SUMS.part
6ef2d9436caa550bec7defda1d769bffc2e0fe8f7acf99e2fbcefd2e9b920a64  guix-build-7f73f422a49f/output/x86_64-w64-mingw32/bitcoin-7f73f422a49f-win64-debug.zip
91c8c3f0aec623f4183c4ebc141816dddb4482509e99812d32298f3e83da920f  guix-build-7f73f422a49f/output/x86_64-w64-mingw32/bitcoin-7f73f422a49f-win64-setup-unsigned.exe
1de7cd32d74990cfcd4e00428c53a35742b7a480f18ae1bd60dbf91b60400947  guix-build-7f73f422a49f/output/x86_64-w64-mingw32/bitcoin-7f73f422a49f-win64-unsigned.tar.gz
e4d4ba33bbdd8d9181c8422af31ad6127f3627c327a93cbe8245d7154b5a936c  guix-build-7f73f422a49f/output/x86_64-w64-mingw32/bitcoin-7f73f422a49f-win64.zip
```

fanquake committed:
depends: don't restrict --enable-lto to non-guix cctools

This wasn't actually disabling LTO support anyways, because it's enabled
by default.

fanquake commented:
> Did --enable-lto-support previously fail for guix builds?

I'm not exactly sure, but I wouldn't have thought so, as we would have been using the same clang / llvm-config versions in Guix. It looks like it's just been working silently for a while, as the option defaults to yes, and llvm-config is found even if not specified explicitly.

fanquake commented:
I've rebased to incorporate recent changes to master. However this branch is still not quite working for macOS Guix builds. Our libtool isn't properly recognizing LLVM bitcode, and it's not yet clear to me why that is the case, as cctools (libtool) is definitely being compiled with lto support enabled.

fanquake commented:
Just a dump of what I've looked at so far. Doing builds with the branch in #25391 (which includes this change) for convenience.

Building depends locally with `LTO=1`, i.e `make -C depends/ HOST=x86_64-apple-darwin LTO=1` works as expected, and ultimately produces working binaries, using LTO, after compiling using the depends CONFIG_SITE.

However building the same branch in Guix, i.e `time HOSTS="x86_64-apple-darwin arm64-apple-darwin" ./contrib/guix/guix-build`, bombs out during the depends build with:
```bash
/bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool -static -o libminiupnpc.a miniwget.o minixml.o igd_desc_parse.o minisoap.o miniupnpc.o upnpreplyparse.o upnpcommands.o upnperrors.o connecthostport.o portlistingparse.o receivedata.o upnpdev.o addr_is_reserved.o minissdpc.o
error: /bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool: file: miniwget.o is not an object file (not allowed in a library)
error: /bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool: file: minixml.o is not an object file (not allowed in a library)
error: /bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool: file: igd_desc_parse.o is not an object file (not allowed in a library)
error: /bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool: file: minisoap.o is not an object file (not allowed in a library)
error: /bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool: file: miniupnpc.o is not an object file (not allowed in a library)
error: /bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool: file: upnpreplyparse.o is not an object file (not allowed in a library)
error: /bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool: file: upnpcommands.o is not an object file (not allowed in a library)
error: /bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool: file: upnperrors.o is not an object file (not allowed in a library)
error: /bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool: file: connecthostport.o is not an object file (not allowed in a library)
error: /bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool: file: portlistingparse.o is not an object file (not allowed in a library)
error: /bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool: file: receivedata.o is not an object file (not allowed in a library)
error: /bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool: file: upnpdev.o is not an object file (not allowed in a library)
error: /bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool: file: addr_is_reserved.o is not an object file (not allowed in a library)
error: /bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool: file: minissdpc.o is not an object file (not allowed in a library)
make[1]: *** [Makefile:302: libminiupnpc.a] Error 1
```

Looking at `miniupnpc.o`, the object files are LLVM bitcode:
```bash
file miniupnpc/2.2.2-fcea302a1da/miniupnpc.o
miniupnpc/2.2.2-fcea302a1da/miniupnpc.o: LLVM bitcode, wrapper x86_64
```

and they should be supported if libtool in cctools is compiled with LTO support (`-DLTO_SUPPORT`), which it is:
```bash
/root/.guix-profile/bin/clang -DPACKAGE_NAME=\"cctools\" -DPACKAGE_TARNAME=\"cctools\" -DPACKAGE_VERSION=\"973.0.1\" -DPACKAGE_STRING=\"cctools\ 973.0.1\" -DPACKAGE_BUGREPORT=\"t.poechtrager@gmail.com\" -DPACKAGE_URL=\"\" -DHAVE_STDIO_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_STRINGS_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_UNISTD_H=1 -DSTDC_HEADERS=1 -DHAVE_DLFCN_H=1 -DLT_OBJDIR=\".libs/\" -DEMULATED_HOST_CPU_TYPE=16777223 -DEMULATED_HOST_CPU_SUBTYPE=3 -D__STDC_LIMIT_MACROS=1 -D__STDC_CONSTANT_MACROS=1 -DHAVE_EXECINFO_H=1 -DHAVE_UTIMENSAT=1 -DHAVE_BCOPY=1 -DHAVE_BCMP=1 -DHAVE_BZERO=1 -DHAVE_INDEX=1 -DHAVE_RINDEX=1 -DHAVE_STAT_ST_MTIM=1 -DHAVE_CLOCK_GETTIME=1 -I.   -I/bitcoin/depends/x86_64-apple-darwin/native/include -I../include -I../include/foreign -I../libstuff -Wall -Wno-long-long -Wno-import -Wno-format -Wno-deprecated -Wno-unused-variable -Wno-unused-private-field -Wno-unused-function -Wno-invalid-offsetof -Wno-int-conversion -Wno-char-subscripts -Wno-shift-negative-value -Wno-misleading-indentation -Wno-gnu-folding-constant -DLTO_SUPPORT -D__DARWIN_UNIX03 -D__LITTLE_ENDIAN__=1 -O3 -isystem /usr/local/include -isystem /usr/pkg/include  -DLD64_VERSION_NUM=609 -fblocks -D__arm64__ -c -o libtool.o libtool.c
```

We can further confirm that by checking that the libtool binary has symbols that would only be compiled in under LTO support, i.e [`lto_toc_nsyms`](https://github.com/tpoechtrager/cctools-port/blob/6540086c5e12e9c1649fed524b527d8c1793ddc0/cctools/misc/libtool.c#L253-L258):
```bash
nm -C /bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool | grep lto_
0000000000437480 b lto_create
0000000000437488 b lto_create_local
0000000000437490 b lto_dispose
00000000004374c0 b lto_error
000000000040e558 T lto_free
00000000004374b8 b lto_get_cputype
000000000040e5e8 T lto_get_nlist_64
000000000040e564 T lto_get_nsyms
00000000004374a0 b lto_get_num_symbols
00000000004374a8 b lto_get_sym_attr
00000000004374b0 b lto_get_sym_name
0000000000437498 b lto_get_target
0000000000437470 b lto_handle
0000000000437478 b lto_is_object
000000000040e7a8 T lto_symbol_name
000000000040e570 T lto_toc_symbol
0000000000408230 t save_lto_member_toc_info
```
Looking at the error message, `file: miniupnpc.o is not an object file (not allowed in a library)`, there are a number of places that could be coming from inside libtool:
https://github.com/tpoechtrager/cctools-port/blob/6540086c5e12e9c1649fed524b527d8c1793ddc0/cctools/misc/libtool.c#L1541
https://github.com/tpoechtrager/cctools-port/blob/6540086c5e12e9c1649fed524b527d8c1793ddc0/cctools/misc/libtool.c#L1614
https://github.com/tpoechtrager/cctools-port/blob/6540086c5e12e9c1649fed524b527d8c1793ddc0/cctools/misc/libtool.c#L1632
https://github.com/tpoechtrager/cctools-port/blob/6540086c5e12e9c1649fed524b527d8c1793ddc0/cctools/misc/libtool.c#L1663

theuni commented:
Based on the exact error message, looks to me like it has to be: https://github.com/tpoechtrager/cctools-port/blob/6540086c5e12e9c1649fed524b527d8c1793ddc0/cctools/misc/libtool.c#L1663

Which indicates that the file is EITHER:
- "ofiles[i].file_type == OFILE_UNKNOWN" OR
- "ofiles[i].file_type == OFILE_LLVM_BITCODE" without LTO_SUPPORT being defined.

I see that it's defined along with `-D__arm64__` above. Is it possible that it's not defined for x86_64?

Or, different question, if that's actually the x86_64 build, why is "\_\_arm64\_\_" defined?

fanquake commented:
> Or, different question, if that's actually the x86_64 build, why is "__arm64__" defined?

fanquake commented:
> Or, different question, if that's actually the x86_64 build, why is "arm64" defined?

@theuni thanks for taking a look. I was performing the `x86_64-apple-darwin` Guix build on an arm64 machine. iirc the same failure also happens when Guix building for `x86_64-apple-darwin` on x86_64. I'll make sure that is still the case.

fanquake commented:
>  I'll make sure that is still the case.

Guix building for `x86_64-apple-darwin` on `x86_64` has the same issue. So the `-D__arm64__` should be a red herring.

fanquake committed:
cctools: fixup building with LTO

Use lto.h from clang+llvm not libtapi. The later is older,
and comes bundled with the libtapi repo.

Copy libLTO.so when building with FORCE_USE_SYSTEM_CLANG.

fanquake commented:
Rebased, and added some additional changes. This is now working for macOS Guix builds with LTO.

fanquake commented:
Squash the first two commits together to fix a build issue due to missing lto.h.

hebasto commented on depends/packages/native_clang.mk:
```diff
@@ -16,10 +16,13 @@ endef
   cp -P bin/clang++ $($(package)_staging_prefix_dir)/bin/ && \
   cp bin/dsymutil $($(package)_staging_prefix_dir)/bin/$(host)-dsymutil && \
   cp bin/llvm-config $($(package)_staging_prefix_dir)/bin/ && \
+  cp include/llvm-c/ExternC.h $($(package)_staging_prefix_dir)/include/llvm-c && \
```
9b60690b94ca766f8039a7a7a20dda5fecd176ab

Why this line has been added?

FWIW, I've managed to build for both darwin hosts with `LTO=1` in Guix using the #25391 branch without this line.

fanquake replied:
Because it's included by the `lto.h` header.

fanquake replied:
It's not completely suprising builds succeed because the file is only copied over (the existing ExternC.h) inside cctools if it exists.

theuni commented:
Concept ACK. Definitely makes more sense to be taking headers from llvm/cctools than tapi.

hebasto reviewed (approved):
ACK 7f73f422a49f5a85c71f645721be5afc045a6dee

hebasto commented:
Guix builds on `x86_64`:
```
af6e0b5fd494f714a84716161352aee415700c1f50f1f08665a474dc2ed56bad  guix-build-7f73f422a49f/output/aarch64-linux-gnu/SHA256SUMS.part
1ecdbf2b579204cf2086001df3887da2ba50cd3f7b8f735113cce8f1b634fdb2  guix-build-7f73f422a49f/output/aarch64-linux-gnu/bitcoin-7f73f422a49f-aarch64-linux-gnu-debug.tar.gz
878ea08855dc559e62fd1932f8a54d53fc63e0ce7d587345c7326892622feee9  guix-build-7f73f422a49f/output/aarch64-linux-gnu/bitcoin-7f73f422a49f-aarch64-linux-gnu.tar.gz
4d1d35bd329935bc74b82a78a7fb8dd39c3268d4bbb33ecd877079e49de97a88  guix-build-7f73f422a49f/output/arm-linux-gnueabihf/SHA256SUMS.part
c0c5418dc270d716b354d2190fddde3d25794758dee45c9068e9f5aac20e5617  guix-build-7f73f422a49f/output/arm-linux-gnueabihf/bitcoin-7f73f422a49f-arm-linux-gnueabihf-debug.tar.gz
5e13ac8cacf1042746eef7e85695469c50d6afc06dfdb5da53039175e5185b52  guix-build-7f73f422a49f/output/arm-linux-gnueabihf/bitcoin-7f73f422a49f-arm-linux-gnueabihf.tar.gz
218b2463a07503e77711fcf0245bc642eb9ed76e4a3e8b61d66a5c76bdc25c62  guix-build-7f73f422a49f/output/arm64-apple-darwin/SHA256SUMS.part
dd15af2b7a333f3c4b3f75abd307e1789a1629728a0ae33d8a509478ace91134  guix-build-7f73f422a49f/output/arm64-apple-darwin/bitcoin-7f73f422a49f-arm64-apple-darwin-unsigned.dmg
5cdfd9fc347007890991c68746b1bb36e0cced57fee685337b3d079d498f687d  guix-build-7f73f422a49f/output/arm64-apple-darwin/bitcoin-7f73f422a49f-arm64-apple-darwin-unsigned.tar.gz
80f957e78177927ace323ea90d4bad2e8d2b846b199a0dcb085c3ceae8516631  guix-build-7f73f422a49f/output/arm64-apple-darwin/bitcoin-7f73f422a49f-arm64-apple-darwin.tar.gz
3cb9cb04db3ebf328bfabfda874fce09764f3840c98bd850d85f71892617fb99  guix-build-7f73f422a49f/output/dist-archive/bitcoin-7f73f422a49f.tar.gz
e4e7c9d5366b658fc420c1fd3cf12bfb9863b8061c83b06fe2492e934bb75ae9  guix-build-7f73f422a49f/output/powerpc64-linux-gnu/SHA256SUMS.part
da4ea7dd9a131901ff2f7fcd6eba573fe946ea671ae6ada418a4ceebcc4bd3b6  guix-build-7f73f422a49f/output/powerpc64-linux-gnu/bitcoin-7f73f422a49f-powerpc64-linux-gnu-debug.tar.gz
221f6e1d5193207eed400c186c8f83747fbf90c3244cd044b36acadbed0c83a6  guix-build-7f73f422a49f/output/powerpc64-linux-gnu/bitcoin-7f73f422a49f-powerpc64-linux-gnu.tar.gz
12d36079b330ec75209219ce58ca580ca5a5eeac64fb7ae424c64e7f3d466ee5  guix-build-7f73f422a49f/output/powerpc64le-linux-gnu/SHA256SUMS.part
3cef0194569d6fa968e92be1d2189be4f1337272dc213c25b8d4afe6b174310c  guix-build-7f73f422a49f/output/powerpc64le-linux-gnu/bitcoin-7f73f422a49f-powerpc64le-linux-gnu-debug.tar.gz
a8ff109153926b3240f830c3501977046b6c470b67380a01c55aea2401a1a833  guix-build-7f73f422a49f/output/powerpc64le-linux-gnu/bitcoin-7f73f422a49f-powerpc64le-linux-gnu.tar.gz
70b3a841bb232fefe27f857ee826ea6a84f59bc781acda608bd6d9385a1ef149  guix-build-7f73f422a49f/output/riscv64-linux-gnu/SHA256SUMS.part
841a276263877027f256c3b7ad1c5bc3ae3ad4d6a7157690daa947f6809776e2  guix-build-7f73f422a49f/output/riscv64-linux-gnu/bitcoin-7f73f422a49f-riscv64-linux-gnu-debug.tar.gz
6cce2c36c2a53035d7ee5d285c8b4b99256772c7a9dca1739f0639ecc0bed71d  guix-build-7f73f422a49f/output/riscv64-linux-gnu/bitcoin-7f73f422a49f-riscv64-linux-gnu.tar.gz
e910b32efebca682cf8bc3994d2c7513b16cdeb6f94cef906efb6629ec53268d  guix-build-7f73f422a49f/output/x86_64-apple-darwin/SHA256SUMS.part
3a4bc1573d9c576184d69b953ea6a85650d4e15b71503b909ccd4fb8a69ca383  guix-build-7f73f422a49f/output/x86_64-apple-darwin/bitcoin-7f73f422a49f-x86_64-apple-darwin-unsigned.dmg
60a114ece1d9a274d9f3d80a9633eb22d589213ba53a956f3b1138d7c8f55560  guix-build-7f73f422a49f/output/x86_64-apple-darwin/bitcoin-7f73f422a49f-x86_64-apple-darwin-unsigned.tar.gz
b6c6eca3e81bfadfed3ddcb7830857c954c8a7b0ff458f14050f4952240ebf27  guix-build-7f73f422a49f/output/x86_64-apple-darwin/bitcoin-7f73f422a49f-x86_64-apple-darwin.tar.gz
aab972c6f051c379a20eb694128a4df4e3737758fbca3902d28d72cd93e6d33a  guix-build-7f73f422a49f/output/x86_64-linux-gnu/SHA256SUMS.part
b8e643eeb72cf04664462f815497cbef1443f2e113de14bc93b058d83836c7df  guix-build-7f73f422a49f/output/x86_64-linux-gnu/bitcoin-7f73f422a49f-x86_64-linux-gnu-debug.tar.gz
3b40fb5f1341cc3b49c9eccc68fc75e1cc7e7d2c6a5664fa2f8f095290e3733a  guix-build-7f73f422a49f/output/x86_64-linux-gnu/bitcoin-7f73f422a49f-x86_64-linux-gnu.tar.gz
96138d5f591a8ba8589eb81d2c79ee269556996ae93c6a3fc77746e101b60cab  guix-build-7f73f422a49f/output/x86_64-w64-mingw32/SHA256SUMS.part
6ef2d9436caa550bec7defda1d769bffc2e0fe8f7acf99e2fbcefd2e9b920a64  guix-build-7f73f422a49f/output/x86_64-w64-mingw32/bitcoin-7f73f422a49f-win64-debug.zip
91c8c3f0aec623f4183c4ebc141816dddb4482509e99812d32298f3e83da920f  guix-build-7f73f422a49f/output/x86_64-w64-mingw32/bitcoin-7f73f422a49f-win64-setup-unsigned.exe
1de7cd32d74990cfcd4e00428c53a35742b7a480f18ae1bd60dbf91b60400947  guix-build-7f73f422a49f/output/x86_64-w64-mingw32/bitcoin-7f73f422a49f-win64-unsigned.tar.gz
e4d4ba33bbdd8d9181c8422af31ad6127f3627c327a93cbe8245d7154b5a936c  guix-build-7f73f422a49f/output/x86_64-w64-mingw32/bitcoin-7f73f422a49f-win64.zip
```
